New resource/data source: `azurerm_mssql_database`

yupwei68 opened the pull request:
fixes terraform-providers/terraform-provider-azurerm#2849
fixes terraform-providers/terraform-provider-azurerm#3978
fixes terraform-providers/terraform-provider-azurerm#1126
addresses #4185 

![mssql db pass](https://user-images.githubusercontent.com/56525716/76484262-4243ee00-6454-11ea-8b1a-c55e19423e66.png)

jackofallops commented on azurerm/internal/services/mssql/data_source_mssql_database.go:
```diff
@@ -0,0 +1,67 @@
+	resp, err := client.Get(ctx, serverId.ResourceGroup, serverId.Name, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Error: Database %q (Resource Group %q, SQL Server %q) was not found", name, serverId.ResourceGroup, serverId.Name)
```
```suggestion
			return fmt.Errorf("Database %q (Resource Group %q, SQL Server %q) was not found", name, serverId.ResourceGroup, serverId.Name)
```

jackofallops commented on azurerm/internal/services/mssql/data_source_mssql_database.go:
```diff
@@ -0,0 +1,67 @@
+			return fmt.Errorf("Error: Database %q (Resource Group %q, SQL Server %q) was not found", name, serverId.ResourceGroup, serverId.Name)
+		}
+
+		return fmt.Errorf("Error making Read request on AzureRM Database %s (Resource Group %q, SQL Server %q): %+v", name, serverId.ResourceGroup, serverId.Name, err)
```
```suggestion
		return fmt.Errorf("Failure making Read request on AzureRM Database %s (Resource Group %q, SQL Server %q): %+v", name, serverId.ResourceGroup, serverId.Name, err)
```

jackofallops commented on azurerm/internal/services/mssql/data_source_mssql_database.go:
```diff
@@ -0,0 +1,67 @@
+				Required: true,
+			},
+
+			"mssql_server_id": {
```
This could possibly be misleading as there is not an `azurerm_mssql_server` resource (yet)?

yupwei68 commented:
Hi @jackofallops Thanks for your comments. Corresponding changes have been pushed. Please continue to review the PR while you're available. Tests have passed.
![mssql db pass2](https://user-images.githubusercontent.com/56525716/76931692-3361ad80-6924-11ea-914d-b36b940d1909.png)
![datasource mssql db pass](https://user-images.githubusercontent.com/56525716/76931709-39f02500-6924-11ea-9a7a-902167ca4aa7.png)

jackofallops commented on azurerm/internal/services/mssql/validate/mssql_database_test.go:
```diff
@@ -0,0 +1,139 @@
+			valid: true,
+		},
+	}
+	var validationFunction = ValidateMsSqlDBSkuName()
```
```suggestion
	var validationFunction = MsSqlDBSkuName()
```

jackofallops commented on azurerm/internal/services/mssql/validate/mssql_elastic_pool.go:
```diff
@@ -0,0 +1,24 @@
+		return warnings, errors
+	}
+
+	if _, _, _, err := parse.ParseArmMSSqlElasticPoolId(v); err != nil {
```
```suggestion
	if _, _, _, err := parse.MSSqlElasticPoolId(v); err != nil {
```

jackofallops commented on website/docs/d/mssql_database.html.markdown:
```diff
@@ -0,0 +1,56 @@
+
+* `sql_server_id` - The id of the Ms SQL Server on which to create the database.
+
+## Argument Reference
```
```suggestion
## Attribute Reference
```

jackofallops commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,110 @@
+
+* `sample_name` - (Optional) Specifies the name of the sample schema to apply when creating this database. Possible value is `AdventureWorksLT`.
+
+* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` , and etc.
```
We should probably have a note here about the default value, since I believe it may be [different between regions](https://azure.microsoft.com/en-gb/updates/azure-sql-database-default-configuration-changing-soon/)? 
```suggestion
* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2`. 
~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```

jackofallops commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,110 @@
+
+* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` , and etc.
+
+* `source_database_id` - (Required) The id of the source database to be copied to create the new database.
```
```suggestion
* `source_database_id` - (Optional) The id of the source database to be copied to create the new database. This should only be used for databases with `create_mode` values that use another database as reference.
```

jackofallops commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,110 @@
+
+* `name` - (Required) The name of the Ms SQL Database. Changing this forces a new resource to be created.
+
+* `sql_server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
```
As mentioned above, possibly should be renamed to avoid confusion in future?  Also, possibly add a note to avoid confusion for "Serverless" still needing a Server resource to be associated to?
```suggestion
* `sql_server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
~> **NOTE:** This setting is still required for "Serverless" SKU's
```

jackofallops commented on azurerm/internal/services/mssql/resource_arm_mssql_database.go:
```diff
@@ -0,0 +1,398 @@
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
```
This may need to be longer. The older database resource had 60 minutes, and I suspect that was due to some create_option values may take some time to complete?

yupwei68 commented:
Hi @jackofallops , thanks for comments with your experience. Corresponding changes are made. Not `forceNew` fields are tested in updates.  I have corrected my regular expression for `skuName` to include all the sku name available. I have added some checks like `source_database_id ` and `restore_point_in_time` in some `create_modes`. And as for your example, `min_capacity` and `auto_pause_delay` will be ignored in API in  other than `GP_S` databases, so I omits this check. Other checks like `read_replica_count` (settable under HS), `read_scale`, `zone_redundant`(settable under BC and Premium), they will be assigned default values in creation in all sku types, so during `terraform plan -refresh=false`, they might raise unexpected errors if the check of whether the fields are assigned  values under an unsuitable DB is added. So I prefer to leave this check to API during `terraform apply`. Tests are all passed. Please continue reviewing while you're available.

yupwei68 commented:
Hi @jackofallops Sorry for the errors made. And thanks for your comments. Corresponding changes are made. Please continue your review.

jackofallops commented on azurerm/internal/services/mssql/resource_arm_mssql_database.go:
```diff
@@ -0,0 +1,401 @@
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.IntBetween(1, 1024),
```
I think these values may be too restrictive?  Different vCore / DTU values have different limits, so I think a basic check is still the best option here, but update the docs to point the user at documentation resources for the value, such as: https://docs.microsoft.com/en-us/azure/sql-database/sql-database-service-tiers-general-purpose-business-critical
```suggestion
				ValidateFunc: validation.IntBetween(1, 4096),
```

Note: we may need to change it to `IntAtLeast(1)` in future if the service increases the upper boundary.

jackofallops commented on azurerm/internal/services/mssql/resource_arm_mssql_database.go:
```diff
@@ -0,0 +1,401 @@
+	serverClient := meta.(*clients.Client).MSSQL.ServersClient
+	serverResp, err := serverClient.Get(ctx, serverId.ResourceGroup, serverId.Name)
+	if err != nil {
+		return fmt.Errorf("Failure in making Read request on MsSql Server  %q (Resource Group %q): %s", serverId.Name, serverId.ResourceGroup, err)
```
whitespace
```suggestion
		return fmt.Errorf("Failure in making Read request on MsSql Server %q (Resource Group %q): %s", serverId.Name, serverId.ResourceGroup, err)
```

jackofallops commented on azurerm/internal/services/mssql/validate/mssql_database.go:
```diff
@@ -0,0 +1,75 @@
+	min := 60
+	max := 10080
+	if (v < min || v > max) && v%10 != 0 && v != -1 {
+		errors = append(errors, fmt.Errorf("expected %s to be in the range (%d - %d) and divisible by 10 or -1, got %d", k, min, max, v))
```
This shouldn't be able to be `-1`?
```suggestion
		errors = append(errors, fmt.Errorf("expected %s to be in the range (%d - %d) and divisible by 10, got %d", k, min, max, v))
```

yupwei68 replied:
In the API, they suppose `-1` as disable `auto pause delay`. So I explain it in the resource doc:`A value of -1 means that automatic pause is disabled`.

jackofallops commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,112 @@
+* `name` - (Required) The name of the Ms SQL Database. Changing this forces a new resource to be created.
+
+* `server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
+~> **NOTE:** This setting is still required for "Serverless" SKU's
```
markdown formatting
```suggestion
* `server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.

~> **NOTE:** This setting is still required for "Serverless" SKU's
```

jackofallops commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,112 @@
+* `sample_name` - (Optional) Specifies the name of the sample schema to apply when creating this database. Possible value is `AdventureWorksLT`.
+
+* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_S_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` ,`DW100c`, `DS100`.
+~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```
markdown formatting
```suggestion
* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_S_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` ,`DW100c`, `DS100`.

~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```

yupwei68 commented:
Hi @jackofallops Thanks for the comments. Corresponding changes are updated, except one unresolved comment. Looking forward to get merged, too.

katbyte commented on azurerm/internal/services/mssql/tests/resource_arm_mssql_database_test.go:
```diff
@@ -0,0 +1,634 @@
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-mssqldb-%[1]d"
```
can we shorten this to just the service abbreviation 
```suggestion
  name     = "acctestRG-mssql-%[1]d"
```

katbyte commented on website/docs/r/mssql_database.html.markdown:
```diff
@@ -0,0 +1,114 @@
+
+~> **NOTE:** This setting is still required for "Serverless" SKU's
+
+* `auto_pause_delay_in_minutes` - (Optional) Time in minutes after which database is automatically paused. A value of -1 means that automatic pause is disabled. This property is only settable for General Purpose Serverless databases.
```
could we also include it needs to be divisible by 10?
```suggestion
* `auto_pause_delay_in_minutes` - (Optional) Time in minutes after which database is automatically paused. A value of `-1` means that automatic pause is disabled. This property is only settable for General Purpose Serverless databases.
```

katbyte commented on azurerm/internal/services/mssql/validate/mssql_database.go:
```diff
@@ -0,0 +1,75 @@
+	return warnings, errors
+}
+
+func MsSqlDBMinCapacity(i interface{}, k string) (warnings []string, errors []error) {
```
could we just write a `validate.FloatInSlice` and put it into helpers for this instead of a custom validation function?

yupwei68 commented:
Hi @katbyte , Thanks for your review comments. Corresponding changes are updated. All tests have been retested with pass.

katbyte commented on azurerm/internal/services/mssql/resource_arm_mssql_database.go:
```diff
@@ -0,0 +1,363 @@
+					string(sql.DatabaseReadScaleEnabled),
+					string(sql.DatabaseReadScaleDisabled),
+				}, false),
+			},
```
Given this is a boolean enabled/disabled property could we make it TypeBool?
```suggestion
			"read_scale_enabled": {
				Type:     schema.TypeBool,
				Optional: true,
				Computed: true,
			},
```

katbyte commented on azurerm/internal/services/mssql/resource_arm_mssql_database.go:
```diff
@@ -0,0 +1,363 @@
+				DiffSuppressFunc: suppress.CaseDifference,
+			},
+
+			"source_database_id": {
```
Would this make more sense as
```suggestion
			"creation_source_database_id": {
```

katbyte commented on website/docs/d/mssql_database.html.markdown:
```diff
@@ -0,0 +1,56 @@
+
+* `collation` - The collation of the database. 
+
+* `elastic_pool_id` - The id of the elastic pool containing this database.
```
```suggestion
* `elastic_pool_id` - The ID of the elastic pool containing this database.
```

katbyte commented on website/docs/d/mssql_database.html.markdown:
```diff
@@ -0,0 +1,56 @@
+
+* `elastic_pool_id` - The id of the elastic pool containing this database.
+
+* `license_type` - The license type to apply for this database.
```
```suggestion
* `license_type` - The license type applied to this database.
```

katbyte commented on website/docs/d/mssql_database.html.markdown:
```diff
@@ -0,0 +1,56 @@
+
+* `read_scale` - If enabled, connections that have application intent set to readonly in their connection string may be routed to a readonly secondary replica.
+
+* `sku_name` - The name of the sku of the database.
```
```suggestion
* `sku_name` - The name of the sku used by the database.
```

yupwei68 commented:
Hi @katbyte ,Sure! Thanks. Please continue your review.
